packages/custom-elements/src/CustomElementState.ts

```diff
@@ -0,0 +1,19 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt The complete set of authors may be found
+ * at http://polymer.github.io/AUTHORS.txt The complete set of contributors may
+ * be found at http://polymer.github.io/CONTRIBUTORS.txt Code distributed by
+ * Google as part of the polymer project is also subject to an additional IP
+ * rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+export const CustomElementState = {
+  custom: 1,
+  failed: 2,
+};
+export type CustomElementState =
+    typeof CustomElementState.custom|typeof CustomElementState.failed;
```

Comment:
With this code:

```typescript
const CustomElementState = {
  custom: 1,
  failed: 2,
};
```

The inferred type of `CustomElementState` is `{custom: number, failed: number}`. If we add `as const` onto the end, the inferred type is: `{readonly custom: 1, readonly failed: 2}`

So it means that the specific values are part of the type, and also they should not be reassigned later.

Reply:
I still don't understand why the current types imply that error but I tried changing an assignment to an element's `__CE_state` to a different number and it complained with this error: `error TS2322: Type '1234' is not assignable to type '1 | 2 | undefined'.`, which is the effect I was hoping for.

I'll need to go and learn more about why TS's type system considers this restricted in this way here - maybe the implied `readonly`-ness is the key part?

Reply:
(kinda continuing from offline) I read some of [the release notes for `as const`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) and found this part:

> no literal types in that expression should be widened (e.g. no going from "hello" to string)

I didn't pick up that difference when I originally read your last comment but now I think I understand what's happening: the `as const` causes the type of the `custom` to be the single value `1` rather than `number`, which affects `typeof CustomElementState.custom`, etc. (and same for `failed`)